### PR TITLE
replace wrapped NodeId with native discv5 type (except for caveat in description)

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?
Fix https://github.com/ethereum/trin/issues/539 

### How was it fixed?
Update discv5 to ``v0.3.0``

Use discv5 node_id type and remove our encapsulated one

#### Known issue
We can't fully remove it till we add ``#[serde(with = "SerHex::<StrictPfx>")]`` to ``sigp/enr`` it is used for recursive query tracing. It is a issue with serialization with hashmaps, if we don't have this flag checked.

https://github.com/sigp/enr/pull/34

^ here is a PR I made to resolve it.

I think we can still merge this even though we still have this 1 reliance, since our goal is to only have to use ``discv5::enr::NodeId`` and not maintain our own and this is one step closer. In the next version of discv5 we should be able to fully remove our wrapper.
### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

### CODE REVIEW

File: trin-types/src/node_id.rs
Comment (ogenev) [inline trin-types/src/node_id.rs:10]: Let's implement [Deref](https://doc.rust-lang.org/std/ops/trait.Deref.html) trait instead of explicitly implementing those methods.

File: trin-types/src/node_id.rs
Comment (ogenev) [inline trin-types/src/node_id.rs:8]:
```suggestion
pub struct NodeId(pub discv5::enr::NodeId);
```

File: trin-types/src/node_id.rs
Comment (ogenev) [inline trin-types/src/node_id.rs:1]: It is also useful to implement `From<discv5::enr::NodeId) for NodeId` so we can convert between these types with ease.

**Review by ogenev — COMMENTED** [review]
Let's wait for https://github.com/ethereum/trin/pull/666 to land first as there is some overlap with the `NodeId` move to `trin_type`

**Note:** This is a temporal solution until discv5 updates their enr  crate version to include this update: https://github.com/sigp/enr/pull/27. Then we can use the native discv5 NodeId without a wrapper.

File: trin-types/src/node_id.rs
Comment (KolbyML) [reply trin-types/src/node_id.rs:1]: image
Seems like my editor isn't happy with that

File: trin-types/src/node_id.rs
Comment (KolbyML) [reply trin-types/src/node_id.rs:10]: image
image
How would that work with Rust? Sorry I must not be familiar enough with how Deref works, but that is what happens when I attempt to use this for example

File: trin-types/src/node_id.rs
Comment (KolbyML) [reply trin-types/src/node_id.rs]: resolved

File: trin-types/src/node_id.rs
Comment (ogenev) [reply trin-types/src/node_id.rs:10]: On line 30 you want to return `Self` which is the wrapped structure NodeId, so it should be something like:

```rust          Ok(NodeId(discv5::enr::NodeId::parse(result.as_slice()).expect("Could not parse decoded NodeID")))
```

File: trin-types/src/node_id.rs
Comment (KolbyML) [reply trin-types/src/node_id.rs:1]: Ty for the pointers and the help resolved

**Comment by KolbyML** [comment]
@ogenev ok your issue's have been resolved

**Comment by ogenev** [comment]
> @ogenev ok your issue's have been resolved

Need to rebase this before the final review.

File: trin-history/src/jsonrpc.rs
Comment (ogenev) [inline trin-history/src/jsonrpc.rs:174]: I think we can remove this conversion and use `node_id.into()` on the next.

Comment (KolbyML) [reply trin-history/src/jsonrpc.rs]: solved

**Comment by ogenev** [comment]
@njgheorghita I think we used some node id hacks to convert between values, could you remind me where those are, and should we update them in this PR?

File: trin-types/src/node_id.rs
Comment (njgheorghita) [inline trin-types/src/node_id.rs:55]: We should return an error here instead of panicking with `expect`

File: trin-types/src/node_id.rs
Comment (njgheorghita) [inline trin-types/src/node_id.rs:14]: I prefer to see all type imports at the top of the file. I'd go with something like `use discv5::enr::NodeId as Discv5NodeId;`.  However, an exception would be something like line 53, where I think it's fine to have an in-line import inside `map_err(..)`

**Review by njgheorghita — CHANGES_REQUESTED** [review]
@ogenev I'm not sure I remember what kind of hacks you're referencing...

Imo, the purpose of this issue should really be to consolidate all of our usage of `NodeId` into a single type. This will make the transition to using discv5's native `NodeId` much simpler once it implements serde natively. @KolbyML If you can go through the repo and convert all usage of `NodeId` to use `trin_types::node_id::NodeId` then it will be much easier to have a complete understanding of the consequences of the changes implemented to the type in this pr

File: ethportal-peertest/src/scenarios/find.rs
Comment (ogenev) [inline ethportal-peertest/src/scenarios/find.rs:89]: I think if we replace the `NodeId`  type in  `ethportal-api/src/types/query_trace.rs`, we will not need this conversion.

File: ethportal-peertest/src/scenarios/find.rs
Comment (ogenev) [inline ethportal-peertest/src/scenarios/find.rs:94]: Same comment as above,

File: ethportal-peertest/src/scenarios/find.rs
Comment (ogenev) [inline ethportal-peertest/src/scenarios/find.rs:165]: Same comment as above,

File: ethportal-api/src/types/node_id.rs
Comment (ogenev) [inline ethportal-api/src/types/node_id.rs:29]: We don't need this structure anymore, we can use `discv5::enr::NodeId`.

Comment (KolbyML) [reply ethportal-api/src/types/node_id.rs:29]: https://github.com/sigp/enr/pull/34 here I list why we need it only for tracing. Without ``#[serde(with = "SerHex::<StrictPfx>")`` ``json!(Hashmap<NodeId, xyz>) will panic. This is used in the query tracing code.

If the concensus is to just wait till the next discv5 till this is resolved, that is fine. I just thought we could do most of the transition.

Comment (KolbyML) [reply ethportal-api/src/types/node_id.rs:29]: I updated my write up with the information as well, https://github.com/ethereum/trin/pull/669#issue-1669551007 . It is unfortunate.

Comment (ogenev) [reply ethportal-api/src/types/node_id.rs:29]: I see, ok, no need to wait for IMO but we need to create an issue after this is merged.

**Review by ogenev — APPROVED** [review]
Looks good :+1:

**Pull request merged**
